manta-util/src/time/lock.rs

```diff
@@ -0,0 +1,196 @@
+    pub const fn get(&self) -> &T {
+        &self.value
+    }
+
+    /// Returns the last [`Instant`] that `self` was modified. See [`elapsed`](Self::elapsed) to get
+    /// the amount of time since the last modification.
+    #[inline]
+    pub const fn modified_at(&self) -> Instant {
+        self.instant
+    }
+
+    /// Returns the amount of time that has elapsed since the last modification of the underlying
+    /// value. See [`modified_at`](Self::modified_at) to get the [`Instant`] of the last
+    /// modification.
+    #[inline]
+    pub fn elapsed(&self) -> Duration {
+        self.instant.elapsed()
+    }
+
+    /// Returns `true` if the amount of time elapsed since the last modification is larger than the
+    /// `timeout`.
+    #[inline]
+    pub fn has_expired(&self, timeout: Duration) -> bool {
+        self.elapsed() >= timeout
+    }
+
+    /// Resets the modification time to the value returned by [`Instant::now`].
+    #[inline]
+    pub fn tap(&mut self) {
+        self.instant = Instant::now();
+    }
+
+    /// Sets the internal value to `value` returning the old value.
+    #[inline]
+    pub fn set(&mut self, value: T) -> T {
+        self.mutate(move |t| mem::replace(t, value))
+    }
+
+    /// Mutates the internal value using `f`, resetting the modification time to [`Instant::now`].
+    #[inline]
+    pub fn mutate<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut T) -> R,
+    {
+        self.mutate_timed(move |value, _| f(value))
+    }
+
+    /// Mutates the internal value with the [`Instant`] of the last modification to `self`,
+    /// resetting the modification time to [`Instant::now`].
+    #[inline]
+    pub fn mutate_timed<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut T, Instant) -> R,
+    {
+        let result = f(&mut self.value, self.instant);
+        self.tap();
+        result
+    }
+
+    /// Mutates the internal value using `f` if the amount of time since the last modification to
+    /// `self` was longer than `timeout`.
+    #[inline]
+    pub fn mutate_if_expired<F, R>(&mut self, timeout: Duration, f: F) -> Option<R>
+    where
+        F: FnOnce(&mut T) -> R,
+    {
+        if self.has_expired(timeout) {
+            Some(self.mutate(f))
+        } else {
+            None
+        }
+    }
+
+    /// Sets the internal value to `value` if the amount of time since the last modification to
+    /// `self` was longer than `timeout`.
+    #[inline]
+    pub fn set_if_expired(&mut self, timeout: Duration, value: T) -> Option<T> {
+        self.set_with_if_expired(timeout, move || value)
+    }
+
+    /// Sets the internal value using `value` if the amount of time since the last modification to
+    /// `self` was longer than `timeout`.
+    #[inline]
+    pub fn set_with_if_expired<F>(&mut self, timeout: Duration, value: F) -> Option<T>
+    where
+        F: FnOnce() -> T,
+    {
+        self.mutate_if_expired(timeout, move |t| mem::replace(t, value()))
+    }
```

Comment:
What's the difference between these two functions?

Reply:
One of them executes the `value` function lazily only if the timer has expired, and the other one accepts the value. See for example the difference between the following:

- https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.resize
- https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.resize_with